### PR TITLE
chore(devenv): direct ref of src style from Storybook

### PR DESCRIPTION
**Pull request by asudoh**
#### Changelog

**Changed**

- A change to make our React Storybook code to directly refer to the style (source) in `packages/components`, so change in `packages/components` will update Storybook live.

#### Testing / Reviewing

Testing should make sure our React dev env (Storybook) is not broken.

### CODE REVIEW

**Review by emyarod — APPROVED** [review]
great to have this integration between the vanilla and react packages. looks good to me!

File: packages/react/.storybook/_container.scss
Comment (joshblack) [inline packages/react/.storybook/_container.scss:5]: This should already resolve to the local copy already because `~carbon-components` points to the hoisted `carbon-components` which is a symlink 👍

**Comment by joshblack** [comment]
Just leaving another note, this should already be the behavior because Yarn will link `carbon-components` directly 👍 If not, let me know!

**Comment by asudoh** [comment]
@joshblack The Sass file in Yarn link requires build to get it up-to-date, which is the exact reason of this PR. Please speak up if I’m missing anything. Thanks!

**Comment by joshblack** [comment]
@asudoh not sure I understand, isn't it pointing to `src` in `~carbon-components/src` and not `scss`?

**Comment by asudoh** [comment]
@joshblack I see, that was the point I was missing - Thank you for clarifying!

**Comment by asudoh** [comment]
Got a second thought - `@import '~carbon-components/src/globals/scss/styles.scss'` gives users a wrong impression that such import is supported for application, too. We have deprecated `src` usage (given keeping it makes change in trasnpiler setting a breaking change) and was planned to remove it from `files` in `v10` (though we didn't make it).

**Comment by joshblack** [comment]
@asudoh on the other hand, the usage that is relative is also a non-standard/unsupported way to use Carbon. It also goes against the grain when it comes to the monorepo setup. If we're okay with that form, what would the difference be here?

**Comment by asudoh** [comment]
@joshblack It’s just that direct path to `src` would give user an impression that it’s internal.

**Comment by joshblack** [comment]
Right, but if they're looking towards this file they will see one of two options:

```
../../components

# or

~carbon-components/src/...
```

Both options are invalid/non-suggested usage patterns, but the first one is antithetical to the monorepo setup.

It would appear that both options are non-viable when it comes to users looking towards this as reference, and as a result it shouldn't be expected that either would work. We could remedy the latter with a comment, and we already point to suggested imports in documentation, so I don't know if we should be worried about this.

**Comment by asudoh** [comment]
I see your point @joshblack that documentation may solve the problem. I also want people to avoid copying the `@import` over to their app, because it works in `9.x` and planning to remove it in next major. I'm just an extra step cautious.

**Comment by joshblack** [comment]
> I also want people to avoid copying the @import over to their app, because it works in 9.x and planning to remove it in next major.

Does our documentation suggest to do this? Or is this the only source for how to do this? If it's the former, we should change it to the recommended way. If it's the latter, then we should make a ticket to create this reference. I don't think we should put as much emphasis on people emulating internal dev environment stuff here if we are making an effort to document these things.

**Comment by asudoh** [comment]
@joshblack It's more about our consumers tend to dig into anywhere in our resources to get a sense on what to do. Especially for people speaking English as their second language, source code often works better than documents, as most open-source projects don't have luxury to get their documents translated.

**Comment by joshblack** [comment]
@asudoh I'm confused though, neither of these approaches would help them if they are looking to figure out how to include the styles in their project so why would it matter?

**Comment by asudoh** [comment]
@joshblack If it's relative path people tends to try to follow the path and can figure out where they want to put Carbon Sass code in their project and adjust the path as such.

**Comment by joshblack** [comment]
@asudoh that's making an assumption though, right? Similarly, if we offer docs on how to include styles shouldn't they be referencing that anyways? Or trying to reverse-engineer from there if the reference isn't localized? Wouldn't the local import imply that the dependency was in-fact included in `carbon-components-react` and that they wouldn't have to install it? It just seems like there are too many things that could go wrong in this situation if we are relying on folks to figure things out in this fashion.

**Comment by asudoh** [comment]
@joshblack We cannot make an assumption wrt where people's reverse-engineering effort lands on. I wouldn't get an impression of local import because it's easy to see the path resolved to a different directory under `package`. But more important point here is that we are not relying on usage of `.storybook/webpack.config.js` as a reference. Instead, this change is _avoiding_ usage of `.storybook/webpack.config.js` as a reference, especially preventing `@import '~carbon-components/src/globals/scss/styles.scss'` code being copied over.

**Comment by joshblack** [comment]
> We cannot make an assumption wrt where people's reverse-engineering effort lands on.

Similarly, we cannot plan for where these efforts may land as this is un-bounded. We can only offer a happy path and point people towards that path. Dealing with where people may end up would result in us trying to cover any number of situations.

> I wouldn't get an impression of local import because it's easy to see the path resolved to a different directory under package.

I think it's fair to suggest that it is likely that someone looking at this code:

```scss
@import '../../components/src/globals/scss/styles.scss';
```

May try and find the directory package and end up with something like:

```scss
@import 'carbon-components/src/globals/scss/styles.scss';
```

Which is also incorrect usage. If both options lead to incorrect usage patterns, it seems fair to suggest we should use the pattern that is inline with intended usage in a monorepo.

> But more important point here is that we are not relying on usage of .storybook/webpack.config.js as a reference

Not quite sure I understand this point, is this because of the usage of `~` along with webpack? Would this rely on webpack config or just an understanding of webpack? This also feels like a point towards how we cannot cover all aspects of someone looking at internals.

**Comment by asudoh** [comment]
Probably a better way to describe my intent in this PR is that I'm trying to avoid writing WebPack tilde path that only works in our source, given WebPack tilde path gives people an impress that it should just work.

**Comment by joshblack** [comment]
> Probably a better way to describe my intent in this PR is that I'm trying to avoid writing WebPack tilde path that only works in our source

It seems fair for us to use it because our technology (storybook) is using webpack. It seemed like your concern with this was that folks might not realize that it is using webpack. The response I had to this concern is the thread above 👍

I would also add that there would be other user-groups who may want to use carbon in their storybook and this could serve as a valid reference for that group. I believe this speaks to the point that we shouldn't change how we author internals to suite a use-case or group, as these groups can come from anywhere. Instead we should follow best practices for the tools that we are using and let users ultimately judge whether or not it's applicable to them.

**Comment by asudoh** [comment]
The tilde in import source is only available with WebPack, and thus it's easy for people to see WebPack is in use. The problem is that such tilde path usage gives people an impression that it works out-of-the-box for Carbon (because it starts with `~carbon-components`), but the tilde path in `.storybook/webpack.config.js` works _only_ in Carbon dev env.

**Comment by joshblack** [comment]
> The problem is that such tilde path usage gives people an impression that it works out-of-the-box for Carbon

I think it's important to remember that this can give _some_ people this impression, not all. On the same hand, if someone was looking to mirror our storybook environment then doing things in the proposed way in this PR would **not** work for their use-case.

At this point, which user group should we target? Or can we just say that we use our development tools as intended and leave it up to folks to decide when/where these things are appropriate?

**Comment by asudoh** [comment]
> On the same hand, if someone was looking to mirror our storybook environment then doing things in the proposed way in this PR would not work for their use-case.

You are right on this, and the point is that with the change it's more obvious that people can not just copy the code there.

**Comment by joshblack** [comment]
@asudoh there may be confusion with my last comment so just to clarify I meant using `~` to refer to modules in sass, not the `src` path part 👍 

Regardless, I don't think we should diverge from a preferred pattern to guard against how people may view and use development-only code. We should write code as intended. If we are guarding, then we would need to mirror this philosophy throughout the codebase. I would be hesitant to take on that kind of surface area in the project especially if we are investing time in reference materials.

**Comment by asudoh** [comment]
My point is that usage of tilde import, though it's a common practice for Sass imports in WebPack apps, shouldn't be applied in a way that works only in development environment. `@import '~carbon-components/src/globals/scss/styles.scss'` is not supposed to work for Carbon applications.

**Comment by joshblack** [comment]
> shouldn't be applied in a way that works only in development environment

I think the discussion here is that we should do this because we shouldn't change our coding style to adjust for how folks might look at internals. If that's the case, then we need to embed that perspective into everything we do. Instead, the suggestion is that we should use the tools as-intended and folks can use these resources to help better inform their perspective on how to use things, alongside official support material.

It should also be noted that the relative path option as proposed also is a strategy that only works in a development environment, so it feels like we'd encounter this no matter what.

**Comment by asudoh** [comment]
Another way to look at the tilde import is that the tilde import is for resolving to package location (which gives maintainers/consumers an impression of published package), but in the one in `webpack.config.js` it's to our source code. That's the tricky part for maintainers/consumers and it's more tricky with the path that works only with out source code.

**Comment by joshblack** [comment]
> That's the tricky part for maintainers/consumers and it's more tricky with the path that works only with out source code.

I feel like this statement is just as true with a relative path? e.g. `@import '../../components'`

**Comment by asudoh** [comment]
I see WebPack tilde path gives people looking at the code much bigger impression that the tilde path just works for applications using `carbon-components` package. OTOH relative path gives bigger impression to people that it may be internal.

**Comment by joshblack** [comment]
I can definitely appreciate that distinction 👍 However, usage of something that we've deprecated and since removed isn't really something we should feel on the hook for here as this whole thing is completely development-specific.

**Comment by asudoh** [comment]
People looking for how to achieve Carbon style in our Storybook may land at `.storybook/_container.scss`. At that point, people is highly likely to get an impression that `@import '~carbon-components/src/globals/scss/styles.scss'` just works, due to tilde path usage (and thus doesn't look like dev-specific code).

**Comment by joshblack** [comment]
@asudoh here's what I'm trying to get at as the ideal:

> We should use our tools as expected.

What this means in this context is:

> We should use the `~` import supported by webpack with storybook to refer to a module in our project setup

If we have this concern:

> People looking for how to achieve Carbon style in our Storybook may land at `.storybook/_container.scss`.

Then we should make sure it's documented where to go instead.

If we have this concern:

> At that point, people is highly likely to get an impression that `@import '~carbon-components/src/globals/scss/styles.scss'` just works, due to tilde path usage (and thus doesn't look like dev-specific code).

Then we should comment the code to indicate otherwise.

**Comment by asudoh** [comment]
>> At that point, people is highly likely to get an impression that `@import '~carbon-components/src/globals/scss/styles.scss'` just works, due to tilde path usage (and thus doesn't look like dev-specific code).

> Then we should comment the code to indicate otherwise.

This makes sense, was thinking of it as an alternative. Done in latest commit.

**Review by joshblack — APPROVED** [review]

**Pull request merged**
